source/_posts/2024-06-07-ai-agents-for-the-smart-home.markdown

```diff
@@ -0,0 +1,189 @@
+    action:
+      - service: calendar.get_events
+        target:
+          entity_id: calendar.my_calendar
+        data_template:
+          start_date_time: "{{ today_at('00:00') }}"
+          duration: { "hours": 24 }
+        response_variable: result
+      - stop: ""
+        response_variable: result
+    speech:
+      text: "{{ action_response['calendar.my_calendar'].events | length }} events"
+{% endraw %}
+{% endhighlight %}
+
+We haven't forgotten about custom components either. They can [register their own intents](https://developers.home-assistant.io/docs/intent_handling) or even better: define their own API.
+
+### Custom integrations providing their own LLM APIs
+
+The built-in LLM API is focused on simplicity and being good at the things that it does. The larger the API surface, the easier AI models, especially the smaller ones, can get confused and invoke them incorrectly.
+
+So instead of one large API, we are aiming for many focused APIs. An AI agent will only have access to one API at a time to ensure it has a higher success rate. It will require experimentation to figure out the best API for creating automations, querying the history, and maybe even creating dashboards. When all those APIs are in place, we can start playing with a selector agent that routes incoming requests to the right agent and API.
+
+To find out what APIs work best is a task we need to do as a community. That's why we have designed our API system in a way that [any custom component can provide them](https://developers.home-assistant.io/docs/core/llm/). When configuring an LLM that supports control of Home Assistant, users can pick any of the available APIs.
+
+Custom LLM APIs are written in Python. When a user talks to an LLM, the API is asked to give a collection of tools for the LLM to access, and a partial prompt that will be appended to the user prompt. The partial prompt can provide extra instructions for the LLM on when and how to use the tools.
+
+## Future research
+
+One thing we can do to improve AI in Home Assistant is wait. LLMs, both local and remotely accessible ones, are improving rapidly and new ones are released regularly (fun fact, I started writing this post before GPT4o and Gemini 1.5 were announced). Wait a couple of months and the new Llama, Gemini or GPT release might unlock many new possibilities.
+
+We're going to continue to collaborate with NVIDIA to enable more local AI functionalities. High on our list is making local LLM with function calling easily accessible to all Home Assistant users.
+
+There is also room for us to improve the local models we use. We want to explore fine tuning a model for specific tasks like voice commands or area summarization. This would allow us to get away with much smaller models with better performance and reliability. And the best thing about our community? [People are already working on this <i class="icon-external-link"></i>](https://huggingface.co/acon96/Home-3B-v3-GGUF).
```

Comment:
```suggestion
There is also room for us to improve the local models we use. We want to explore fine-tuning a model for specific tasks like voice commands or area summarization. This would allow us to get away with much smaller models with better performance and reliability. And the best thing about our community? [People are already working on this <i class="icon-external-link"></i>](https://huggingface.co/acon96/Home-3B-v3-GGUF).
```